Normalize hierarchy deletion, refs #13239

jraddaoui opened the pull request:
Avoid nested set update until the end of the full tree deletion to avoid
unnecessary nested set updates.

- Normalize `deleteFullHierarchy()` in QubitObject.
- Check that it's only called from QubitInformationObject and QubitTerm.
- Run hierarchy deletion within transaction in delete description task.
- Use `deleteFullHierarchy()` on term deletion.
- Don't show all descendants on term deletion page.

jraddaoui commented:
See https://projects.artefactual.com/issues/13239#note-5 for a bit more info.

sbreker commented on apps/qubit/modules/term/actions/deleteAction.class.php:
```diff
@@ -45,13 +45,7 @@ public function execute($request)
     {
-      foreach ($this->resource->descendants->andSelf()->orderBy('rgt') as $item)
-      {
-        if (QubitAcl::check($item, 'delete'))
```
Is this ACL check still necessary? It looks like it has been removed.

jraddaoui replied:
Good point @sbreker! Sorry, I forgot to add a note about this ...

The ACL permissions for terms are set at the taxonomy level so, if the user has permission to delete a term, it will also have it to delete its descendants. But, now that you say it, we didn't have that check for IOs either, where the permissions can be set at the IO level. However, I'm not sure if we want to add it as it would be a rare case and it will seriously affect the performance of this hierarchy deletions.

At least this is something @fiver-watson should know, maybe we want to add a note in the documentation if we don't have it already or address it in a different issue.

sbreker reviewed (approved):
Looks great @jraddaoui! 👍

jraddaoui committed:
Normalize hierarchy deletion, refs #13239

Avoid nested set update until the end of the full tree deletion to avoid
unnecessary nested set updates.

- Normalize `deleteFullHierarchy()` in QubitObject.
- Check that it's only called from QubitInformationObject and QubitTerm.
- Run hierarchy deletion within transaction in delete description task.
- Use `deleteFullHierarchy()` on term deletion.
- Don't show all descendants on term deletion page.

Squashed commit of the following:

commit f79265ce67af0c1a2f8b94149178b3fd38ba98b0
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue May 26 20:41:28 2020 +0200

    Don't show all descendants on term deletion page

commit 5462eeac2a3aef768246c02d2df4f453ca07ddf6
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue May 26 20:17:17 2020 +0200

    Use deleteFullHierarchy on term deletion

commit 9a395b7951db00d396b7c4a34ec7a18e9714596a
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue May 26 20:14:14 2020 +0200

    Run hierarchy deletion within transaction in task

commit addadef34d2f0ccfc1410783e7a79ee455fe88eb
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue May 26 19:56:26 2020 +0200

    Check model on deleteFullHierarchy

commit b336c93955b910ebd725382373986d0ddcf4018e
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue May 26 19:39:08 2020 +0200

    Move deleteFullHierarchy to QubitObject
